llvm_ir_dataset_utils/tools/portage_analyze_failures.py

```diff
@@ -0,0 +1,145 @@
+import subprocess
+import re
+import os
+from collections import OrderedDict
+import json
+
+
+def run_equery_depgraph(pkg):
+  pkg = pkg.replace('_', '/', 1)
+  command = f"emerge -pv {pkg}"
+  try:
+    result = subprocess.run(
+        command, shell=True, check=True, capture_output=True, text=True)
+    return result.stdout
+  except subprocess.CalledProcessError as e:
+    if "The following USE changes are necessary to proceed" in e.stderr:
+      print(f"Package {pkg} needs USE flag")
+    use_changes = parse_emerge_output(e.stderr)
+    update_package_use_custom(use_changes)
+    try:
+      result = subprocess.run(
+          command, shell=True, check=True, capture_output=True, text=True)
+      return result.stdout
+    except subprocess.CalledProcessError as e:
+      if "have been masked." in e.stderr:
+        return 1
+      if "The following USE changes are necessary to proceed" in e.stderr:
+        return 2
+      return None
+
+
+def parse_emerge_output(output):
+  use_changes = re.findall(r'>=([^\s]+)\s+([^\n]+)', output)
+  use_changes = [(re.sub(r'-[0-9.]+(?:-r[0-9]+)?$', '', package), flags)
+                 for package, flags in use_changes]
+  return use_changes
+
+
+def update_package_use_custom(use_changes):
+  package_use_dir = '/etc/portage/package.use'
+  custom_file_path = os.path.join(package_use_dir, 'custom')
+
+  if not os.path.exists(package_use_dir):
+    os.makedirs(package_use_dir)
+
+  if not os.path.exists(custom_file_path):
+    open(custom_file_path, 'a').close()
+
+  with open(custom_file_path, 'r+') as f:
+    content = f.read()
+    for package, flags in use_changes:
+      if package not in content:
+        f.write(f"{package} {flags}\n")
+        print(f"Added to {custom_file_path}: {package} {flags}")
+      else:
+        print(f"Package {package} already exists in {custom_file_path}")
+
+
+def parse_depgraph(content):
+  lines = content.split('\n')
+  dependencies = []
+  for line in lines[1:]:
+    match = re.search(r"(\w+-[\w+/-]+?)(?:-\d[\w\._-]*)", line)
+    if match:
+      dependencies.append(match.group(1))
+
+  return dependencies
+
+
+def analyse_neither():
+  failed = []
+  succeeded = []
+  # neither.json is a log analyzing the results of a package whose installation result type is neither.
+  # 'Y' means it has been built successfully.
+  # 'N' means its build failed.
+  with open('./portage-lists/neither.json') as neither_pkgs_files:
```

Comment:
Should these file arguments be CLI flags? Or are they static on all Gentoo installations?

Reply:
In fact, the N/Y analysis here does not exist in checkin. Here is the N/Y that I got from analyzing the build log externally using jupyter-notebook+OpenAI API. maybe I should just delete these analyses.